packages/opentelemetry-context-base/src/types.ts

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** Default type for functions */
+// tslint:disable:no-any
+export type Func<T> = (...args: unknown[]) => T;
+
+/** A key/value object for anything that can be stored in the context */
+export interface ContextStore {
+  [key: string]: unknown;
+}
+
+export interface ContextManager {
```

Comment:
This interface makes it unclear if child contexts should inherit from parent contexts. While that's the behavior of `continuation-local-storage` which is arguably the grandfather of JS context propagation, there is a performance cost of doing so and it's prone to memory leaks.

Not supporting that however makes this interface potentially too flexible, in which case I would probably get/set a single object and not allow individual properties.

Reply:
I thought about this more, and I think think we should go with the second approach where we store a single object. Then there should be a method that runs a callback with the object as its context. This will be much simpler and not ambiguous.

In the future, it will be possible to add an API similar to what is proposed here on top as another project, something like `@opentelemetry/cls` which would allow inheritance and individual properties.

Reply:
I don't know much about the performance of each approach but i would prefer to stick on the behavior of the CLS for the sake of simplicity.
Almost all APM vendors use it so it's well known, changing it might make difficult the process of porting a instrumentation from any vendor to the OpenTelemetry API, which i think is more problematic than already-established performance cost (since the CLS approach is currently used anyway).

Reply:
Do we need `enable()` and `disable()` methods or by default it would be enable immediately after creation?

Reply:
> I don't know much about the performance of each approach but i would prefer to stick on the behavior of the CLS for the sake of simplicity.

I did many benchmarks and not using CLS is faster and safer. It's also easier to reason about to store a single object since it ensures different vendors have full control over how they implement this if they need to have their own implementation. What's also great about the simpler approach is that you can wrap it and build a CLS on top. The opposite is not true.

At Datadog we started with CLS and switched later on for performance and stability reasons. I think most vendors went with CLS because it's the common way to propagate context for applications, but it doesn't mean it's the best way for this type of project, it's just the default.

> Do we need enable() and disable() methods or by default it would be enable immediately after creation?

I would say yes. We have customers who wanted more control over context propagation. It also makes it easier to be used on its own and for integration testing.